web/tests/lib/make_message.js

```diff
@@ -0,0 +1,275 @@
+            create_pm_display_recip(sender_id),
+            create_pm_display_recip(create_next_sender_id()),
+        ],
+        recipient_id: recipient_id ?? create_recipient_id(),
+    };
+}
+
+function make_private_server_msg(opts = {}) {
+    // create private messages
+    const sender_id = opts.sender_id ?? create_next_sender_id();
+    const message_id = opts.id ?? create_next_msg_id();
+
+    const base_props = {
+        edit_history: opts.edit_history,
+        last_edit_timestamp: opts.last_edit_timestamp,
+        submessages: opts.submessages,
+        reactions: opts.reactions,
+        is_me_message: opts.is_me_message,
+        topic_links: opts.topic_links,
+        flags: opts.flags,
+    };
+
+    const sender = {
+        sender_id,
+        avatar_url: opts.avatar_url,
+        client: opts.client,
+        sender_full_name: opts.sender_full_name,
+        sender_email: opts.sender_email,
+        sender_realm_str: opts.sender_realm_str,
+    };
+
+    const pm_details = {
+        display_recipient: opts.display_recipient,
+        sender_id,
+        recipient_id: opts.recipient_id,
+    };
+
+    return {
+        ...msg_base_props(base_props),
+        ...msg_sender_props(sender),
+        ...msg_pm_props(pm_details),
+        id: message_id,
+        content: opts.content ?? "<p>(message content)</p>",
+        content_type: opts.content_type ?? "text/html",
+        timestamp: opts.timestamp ?? create_next_timestamp(),
+        type: "private",
+        local_id: opts.local_id,
+    };
+}
+
+function make_stream_server_msg(opts = {}) {
+    const sender_id = opts.sender_id ?? create_next_sender_id();
+    const message_id = opts.id ?? create_next_msg_id();
+    const stream_id = opts.stream_id ?? create_stream_id();
+
+    const base_props = {
+        edit_history: opts.edit_history,
+        last_edit_timestamp: opts.last_edit_timestamp,
+        submessages: opts.submessages,
+        reactions: opts.reactions,
+        is_me_message: opts.is_me_message,
+        topic_links: opts.topic_links,
+        flags: opts.flags,
+    };
+
+    const sender = {
+        sender_id,
+        avatar_url: opts.avatar_url,
+        client: opts.client,
+        sender_full_name: opts.sender_full_name,
+        sender_email: opts.sender_email,
+        sender_realm_str: opts.sender_realm_str,
+    };
+
+    const stream_details = {
+        stream_id,
+        subject: opts.subject,
+        display_recipient: opts.display_recipient,
+        recipient_id: opts.recipient_id,
+    };
+
+    return {
+        ...msg_base_props(base_props),
+        ...msg_sender_props(sender),
+        ...msg_stream_props(stream_details),
+        id: message_id,
+        content: opts.content ?? "<p>(message content)</p>",
+        content_type: opts.content_type ?? "text/html",
+        timestamp: opts.timestamp ?? create_next_timestamp(),
+        type: "stream",
+        local_id: opts.local_id,
+    };
+}
```

Comment:
I think it'd be valuable for this file to be TypeScript, and return a `Message` or `RawMessage` or whatever the appropriate type is; that really helps to validate your work here.

Reply:
That would be great but that's not an option yet, right? That's my understanding of Anders's comments here:
https://chat.zulip.org/#narrow/stream/6-frontend/topic/typescript.20node.20tests/near/1886768

Reply:
> That would be great but that's not an option yet, right? That's my understanding of Anders's comments here: https://chat.zulip.org/#narrow/stream/6-frontend/topic/typescript.20node.20tests/near/1886768

Yes. That's my understanding of the discussion too.